src/vcall.cpp

```diff
@@ -1326,11 +1326,18 @@ void jitc_var_vcall_collect_data(tsl::robin_map<uint64_t, uint32_t, UInt64Hasher
     if (!it_and_status.second)
         return;
 
-    const Variable *v = jitc_var(index);
+    Variable *v = jitc_var(index);
 
     if (!v->literal && v->stmt && strstr(v->stmt, "self"))
         use_self = true;
 
+    /* Evaluate scalar literals created outside of this virtual function call
+       to avoid baking them inside of the kernel IR. */
+    if (v->literal && v->cse_scope != thread_state(v->backend)->cse_scope) {
```

Comment:
What I had in mind is that -- rather than actually evaluating those literals, we can handle them specially in the part of the code that generates the data array used by the virtual function call. I am referring to the loop at  https://github.com/mitsuba-renderer/drjit-core/blob/master/src/vcall.cpp#L336.

This loop computes the input that is later used to launch a kernel which either

1. Copies a raw pointer value (which is actually a kind of literal) from the CPU into a GPU output array
2. Copies a single data word (1, 2, 4, or 8 bytes) from some region of GPU memory into a GPU output array

What I am suggesting is that we have a 3rd case, which copies any other kind of constant literal (1, 2, 4, or 8 bytes) form CPU memory to the GPU output array. So this would be a sort of generalization of #1.